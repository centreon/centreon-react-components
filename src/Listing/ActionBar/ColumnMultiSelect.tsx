import * as React from 'react';

import { useTranslation } from 'react-i18next';
import { prop } from 'ramda';

import ColumnIcon from '@material-ui/icons/ViewColumn';

import { getVisibleColumns, Props as ListingProps } from '..';
import IconPopoverMultiSelect from '../../InputField/Select/Autocomplete/Multi/IconPopover';
import { labelAddColumns, labelColumns } from '../translatedLabels';
import { SelectEntry } from '../../InputField/Select';
import { Column } from '../models';

type Props = Pick<
  ListingProps<unknown>,
  'columns' | 'columnConfiguration' | 'onSelectColumns'
>;

const toSelectEntries = (columns: Array<Column>): Array<SelectEntry> => {
  return columns.map(({ id, label }) => ({
    id,
    name: label,
  }));
};

const ColumnMultiSelect = ({
  columns,
  columnConfiguration,
  onSelectColumns,
}: Props): JSX.Element => {
  const { t } = useTranslation();

  const visibleColumns = getVisibleColumns({
    columns,
    columnConfiguration,
  });

  const selectColumnIds = (_, updatedColumns) => {
    onSelectColumns?.(updatedColumns.map(prop('id')));
  };

  const resetColumns = () => {
    onSelectColumns?.(columns.map(prop('id')));
  };

  return (
    <IconPopoverMultiSelect
      title={t(labelAddColumns)}
      options={toSelectEntries(columns)}
      value={toSelectEntries(visibleColumns)}
      label={t(labelColumns)}
      onChange={selectColumnIds}
      icon={<ColumnIcon />}
      onReset={resetColumns}
      popperPlacement="bottom-end"
    />
  );
};

export default ColumnMultiSelect;
