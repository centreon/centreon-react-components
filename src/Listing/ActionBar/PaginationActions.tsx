import React from 'react';

import { useTranslation } from 'react-i18next';

import { makeStyles } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import FirstPageIcon from '@material-ui/icons/FirstPage';
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft';
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight';
import LastPageIcon from '@material-ui/icons/LastPage';
import { TablePaginationActionsProps } from '@material-ui/core/TablePagination/TablePaginationActions';

import {
  labelFirstPage,
  labelLastPage,
  labelNextPage,
  labelPreviousPage,
} from '../translatedLabels';

const useStyles = makeStyles((theme) => ({
  root: {
    flexShrink: 0,
    color: theme.palette.text.secondary,
  },
}));

const PaginationActions = ({
  onChangePage,
  page,
  rowsPerPage,
  count,
}: TablePaginationActionsProps): JSX.Element => {
  const classes = useStyles();
  const { t } = useTranslation();

  const changeToFirstPage = (event) => {
    onChangePage(event, 0);
  };

  const changeToPreviousPage = (event) => {
    onChangePage(event, page - 1);
  };

  const changeToNextPage = (event) => {
    onChangePage(event, page + 1);
  };

  const lastPage = Math.ceil(count / rowsPerPage) - 1;

  const isFirstPage = page === 0;
  const isLastPage = page >= lastPage;

  const changeToLastPage = (event) => {
    onChangePage(event, Math.max(0, lastPage));
  };

  return (
    <div className={classes.root}>
      <IconButton
        onClick={changeToFirstPage}
        disabled={isFirstPage}
        aria-label={t(labelFirstPage)}
      >
        <FirstPageIcon />
      </IconButton>
      <IconButton
        onClick={changeToPreviousPage}
        disabled={isFirstPage}
        aria-label={t(labelPreviousPage)}
      >
        <KeyboardArrowLeft />
      </IconButton>
      <IconButton
        onClick={changeToNextPage}
        disabled={isLastPage}
        aria-label={t(labelNextPage)}
      >
        <KeyboardArrowRight />
      </IconButton>
      <IconButton
        onClick={changeToLastPage}
        disabled={isLastPage}
        aria-label={t(labelLastPage)}
      >
        <LastPageIcon />
      </IconButton>
    </div>
  );
};

export default PaginationActions;
