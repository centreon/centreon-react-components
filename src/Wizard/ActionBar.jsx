import React from 'react';

import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';

const useStyles = makeStyles((theme) => ({
  container: {
    position: 'sticky',
    bottom: 0,
    padding: '0px 10px',
    borderTop: `1px solid ${theme.palette.grey[300]}`,
  },
}));

const ActionBar = ({
  disabledNext,
  page,
  isLastPage,
  onCancel,
  onPrevious,
  onNext,
  onFinish,
  labelCancel,
  labelPrevious,
  labelNext,
  labelFinish,
}) => {
  const classes = useStyles();

  const preventEnterKey = (keyEvent) => {
    if ((keyEvent.charCode || keyEvent.keyCode) === 13) {
      keyEvent.preventDefault();
    }
  };

  return (
    <Grid
      container
      direction="row"
      justify="space-between"
      alignItems="center"
      className={classes.container}
    >
      <Grid item>
        {onCancel && (
          <Button
            type="button"
            color="primary"
            onClick={(event) => onCancel(event, 'cancel')}
            onKeyPress={preventEnterKey}
          >
            {labelCancel}
          </Button>
        )}
      </Grid>

      <Grid item>
        {page > 0 && (
          <Button
            type="button"
            color="primary"
            onClick={onPrevious}
            onKeyPress={preventEnterKey}
          >
            {labelPrevious}
          </Button>
        )}

        {isLastPage ? (
          <Button
            type="submit"
            color="primary"
            disabled={disabledNext}
            onClick={onFinish}
            onKeyPress={preventEnterKey}
          >
            {labelFinish}
          </Button>
        ) : (
          <Button
            type="submit"
            color="primary"
            onClick={onNext}
            disabled={disabledNext}
            onKeyPress={preventEnterKey}
          >
            {labelNext}
          </Button>
        )}
      </Grid>
    </Grid>
  );
};

ActionBar.propTypes = {
  disabledNext: PropTypes.bool,
  page: PropTypes.number,
  isLastPage: PropTypes.bool,
  onCancel: PropTypes.func,
  onPrevious: PropTypes.func,
  onNext: PropTypes.func,
  onFinish: PropTypes.func,
  labelCancel: PropTypes.string,
  labelPrevious: PropTypes.string,
  labelNext: PropTypes.string,
  labelFinish: PropTypes.string,
};

ActionBar.defaultProps = {
  disabledNext: false,
  page: 0,
  isLastPage: true,
  onCancel: null,
  onPrevious: null,
  onNext: null,
  onFinish: null,
  labelCancel: 'Cancel',
  labelPrevious: 'Previous',
  labelNext: 'Next',
  labelFinish: 'Finish',
};

export default ActionBar;
