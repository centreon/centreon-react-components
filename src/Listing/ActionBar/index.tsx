import * as React from 'react';

import { useTranslation } from 'react-i18next';

import { makeStyles } from '@material-ui/core';

import { ListingProps } from '../..';
import { labelOf, labelRowsPerPage } from '../translatedLabels';

import StyledPagination from './Pagination';
import PaginationActions from './PaginationActions';
import ColumnMultiSelect from './ColumnMultiSelect';

const useStyles = makeStyles((theme) => ({
  container: {
    display: 'grid',
    gridTemplateColumns: '1fr auto auto',
    width: '100%',
    alignItems: 'center',
    gridGap: theme.spacing(1),
  },
  pagination: {
    padding: 0,
  },
  actions: {
    padding: theme.spacing(1),
  },
}));

type Props = Pick<
  ListingProps<unknown>,
  | 'actions'
  | 'onLimitChange'
  | 'onPaginate'
  | 'paginated'
  | 'currentPage'
  | 'limit'
  | 'totalRows'
  | 'columns'
  | 'columnConfiguration'
  | 'onSelectColumns'
>;

const ListingActionBar = ({
  actions,
  onPaginate,
  onLimitChange,
  paginated,
  totalRows,
  currentPage,
  limit,
  columns,
  columnConfiguration,
  onSelectColumns,
}: Props): JSX.Element => {
  const { t } = useTranslation();
  const classes = useStyles();

  const changeRowPerPage = (event): void => {
    onLimitChange?.(event.target.value);
    onPaginate?.(0);
  };

  const changePage = (_, value: number): void => {
    onPaginate?.(value);
  };

  const labelDisplayedRows = ({ from, to, count }): string =>
    `${from}-${to} ${t(labelOf)} ${count}`;

  return (
    <div className={classes.container}>
      <div className={classes.actions}>{actions}</div>
      {columnConfiguration?.selectable && (
        <ColumnMultiSelect
          columns={columns}
          columnConfiguration={columnConfiguration}
          onSelectColumns={onSelectColumns}
        />
      )}
      {paginated && (
        <StyledPagination
          className={classes.pagination}
          rowsPerPageOptions={[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]}
          labelDisplayedRows={labelDisplayedRows}
          labelRowsPerPage={t(labelRowsPerPage)}
          colSpan={3}
          count={totalRows}
          rowsPerPage={limit}
          page={currentPage}
          SelectProps={{
            native: true,
          }}
          onChangePage={changePage}
          onChangeRowsPerPage={changeRowPerPage}
          ActionsComponent={PaginationActions}
        />
      )}
    </div>
  );
};

export default ListingActionBar;
