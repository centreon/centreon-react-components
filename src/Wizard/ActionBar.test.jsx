import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ActionBar from './ActionBar';

describe('ActionBar', () => {
  it('renders correctly', () => {
    const { container } = render(<ActionBar />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('cancels', () => {
    const mockCancel = jest.fn();

    const { getByText } = render(
      <ActionBar onCancel={mockCancel} labelCancel="Exit" />,
    );

    fireEvent.click(getByText('Exit').parentNode);

    expect(mockCancel).toBeCalled();
  });

  it('cannot finish if form is not valid', () => {
    const { getByText } = render(<ActionBar disabledNext />);

    expect(getByText('Finish').parentNode).toHaveAttribute('disabled');
  });
});
