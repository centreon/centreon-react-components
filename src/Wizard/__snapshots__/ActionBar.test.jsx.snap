// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBar renders correctly 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
>
  <div
    class="MuiGrid-root MuiGrid-item"
  />
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Finish
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
